### PR TITLE
NWT: necessary changes for Brinkman code.

### PR DESCRIPTION
**Pull request by drwells**
These changes are all necessary for the Brinkman code in #418. Part of #546.

### CODE REVIEW

**Comment by drwells** [comment]
Lets move the output stream stuff into a separate function (so that this class is not responsible for writing output files), i.e., `log_torque(Vector3d &foo, Vector3d &barr, std::ostream &out)`.

File: include/ibamr/IBHydrodynamicSurfaceForceEvaluator.h [hunk 1]
```diff
@@ -103,9 +102,40 @@ class IBHydrodynamicSurfaceForceEvaluator
...
+                                                IBTK::Vector3d& viscous_torque,
+                                                const IBTK::Vector3d& X0,
+                                                double time,
+                                                double current_time,
+                                                double new_time);
+    /*!
+     * \brief Set the surface contour value <em> s <\em> of the level set field, such that \f$ \phi(s) = \partial \Omega
+     * \f$.
      */
-    virtual void computeHydrodynamicForce();
```
Comment (drwells) [inline include/ibamr/IBHydrodynamicSurfaceForceEvaluator.h]: We should at least log that we broke this interface under 'incompatible changes' and fix any relevant examples to not use it.

File: include/ibamr/IBHydrodynamicSurfaceForceEvaluator.h [hunk 2]
```diff
@@ -137,7 +167,9 @@ class IBHydrodynamicSurfaceForceEvaluator
      * Fill required patch data and ghost cells.
      */
     void fillPatchData(SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
-                       const double fill_time);
+                       double fill_time,
+                       bool use_cuurent_ctx,
```
Comment (drwells) [inline include/ibamr/IBHydrodynamicSurfaceForceEvaluator.h]:
```suggestion
                       bool use_current_ctx,
```

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 3]
```diff
@@ -70,6 +68,31 @@ sign(const double X)
...
+            side_coord[d] = patch_X_lower[d] + patch_dx[d] * (static_cast<double>(side_idx(d) - patch_lower_idx(d)));
+        }
+        else
+        {
+            side_coord[d] =
+                patch_X_lower[d] + patch_dx[d] * (static_cast<double>(side_idx(d) - patch_lower_idx(d)) + 0.5);
+        }
+    }
+    return;
+} // get_physical_coordinate
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: Do we already have a function that does this somewhere, like in `IndexUtilities`?

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 4]
```diff
@@ -266,6 +336,10 @@ IBHydrodynamicSurfaceForceEvaluator::computeHydrodynamicForce()
                     IBTK::Vector3d n = IBTK::Vector3d::Zero();
                     n(axis) = sign(phi_upper - phi_lower);
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: I think we can use `std::signbit` here.

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 5]
```diff
@@ -336,25 +409,73 @@ IBHydrodynamicSurfaceForceEvaluator::computeHydrodynamicForce()
...
 /////////////////////////////// PROTECTED ////////////////////////////////////
 
 /////////////////////////////// PRIVATE //////////////////////////////////////
 
 void
 IBHydrodynamicSurfaceForceEvaluator::fillPatchData(Pointer<PatchHierarchy<NDIM> > patch_hierarchy,
-                                                   const double fill_time)
+                                                   const double fill_time,
+                                                   bool use_current_ctx,
+                                                   bool use_new_ctx)
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: We should make the arguments to this the same as `computeHydrodynamicForceTorque`: it has the same logic for determining which context to use.

We should also log that this improvement (i.e., we no longer assume that we are at the current time) is incompatible.

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 6]
```diff
@@ -196,29 +219,76 @@ IBHydrodynamicSurfaceForceEvaluator::~IBHydrodynamicSurfaceForceEvaluator()
...
     for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
     {
-        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_ls_solid_idx, integrator_time);
-        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_u_idx, integrator_time);
-        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_p_idx, integrator_time);
-        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_mu_idx, integrator_time);
+        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_ls_solid_idx, time);
+        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_u_idx, time);
+        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_p_idx, time);
+        patch_hierarchy->getPatchLevel(ln)->allocatePatchData(d_mu_idx, time);
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: We don't need to allocate here if the density is constant.

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 7]
```diff
@@ -336,25 +409,73 @@ IBHydrodynamicSurfaceForceEvaluator::computeHydrodynamicForce()
     }
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: Line 408: we don't need to deallocate if the density is constant.

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 8]
```diff
@@ -396,16 +520,23 @@ IBHydrodynamicSurfaceForceEvaluator::fillPatchData(Pointer<PatchHierarchy<NDIM>
             p_vc_ins_hier_integrator->getTransportedViscosityVariable();
         Pointer<CellVariable<NDIM, double> > mu_ins_var = p_vc_ins_hier_integrator->getViscosityVariable();
         RobinBcCoefStrategy<NDIM>* mu_bc_coef = nullptr;
-        int mu_current_idx = -1;
+        int mu_idx = -1;
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]:
```suggestion
        int mu_idx = IBTK::invalid_index;
```

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp [hunk 9]
```diff
@@ -428,55 +559,58 @@ IBHydrodynamicSurfaceForceEvaluator::fillPatchData(Pointer<PatchHierarchy<NDIM>
...
-            hier_p_bdry_fill->setHomogeneousBc(false);
-            hier_p_bdry_fill->fillData(fill_time);
-        }
-        else
-        {
-            TBOX_ERROR(d_object_name << "::IBHydrodynamicSurfaceForceEvaluator():\n"
-                                     << " no valid pressure boundary condition object registered with INS integrator.\n"
-                                     << " This statement should not have been reached");
-        }
+    // Fill ghost cells for pressure
```
Comment (drwells) [inline src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: Note that we need to do this regardless of whether nor not the density is constant, so this fixes a bug.

**Comment by drwells** [comment]
@amneetb and I just read through this together and we are both happy.

@amneetb has test code for a hydrostatic suspended cylinder with zero torque. Otherwise I have only minor comments for things to change. I can handle these little changes.

**Comment by drwells** [comment]
@amneetb we have five examples that still use `computeHydrodynamicForce`:
```
[drwells@learch examples]$ grep -Ri computeHydrodynamicForce
CIB/ex1/example.cpp:            hydro_force->computeHydrodynamicForce(u_idx,
ConstraintIB/eel2d/example.cpp:            hydro_force->computeHydrodynamicForce(u_idx,
ConstraintIB/flow_past_cylinder/example.cpp:            hydro_force->computeHydrodynamicForce(u_idx,
ConstraintIB/moving_plate/example.cpp:            hydro_force->computeHydrodynamicForce(u_idx,
multiphase_flow/ex1/example.cpp:            hydro_force_evaluator->computeHydrodynamicForce();
multiphase_flow/ex2/example.cpp:            hydro_force_evaluator->computeHydrodynamicForce();
```
Do you have a recommended change that preserves the behavior of these examples?

**Comment by drwells** [comment]
(compilation fails because we cannot compile these examples any longer)

**Comment by drwells** [comment]
Clarification: most of these examples don't use this class (I only searched for the function, not class, name). The only examples we need to update are the two multiphase flow examples.

File: src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp (hunk 3 above)
Comment (drwells) [reply src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp]: (I added this to IndexUtilities)

File: include/ibamr/IBHydrodynamicSurfaceForceEvaluator.h [hunk 10]
```diff
@@ -190,12 +222,25 @@ class IBHydrodynamicSurfaceForceEvaluator
     double d_surface_contour_value = 0.0;
 
     /*!
-     * \brief File streams associated for the output.
+     * \brief Whether to write results on a text file.
+     */
+    bool d_write_to_file = false;
+
+    /*!
+     * \brief File streams associated for the output of hydrodynamic force.
```
Comment (drwells) [inline include/ibamr/IBHydrodynamicSurfaceForceEvaluator.h]: As noted above, lets remove this and the other logging in favor of a method taking a `std::ostream &` argument.

File: ibtk/include/ibtk/IndexUtilities.h [hunk 11]
```diff
@@ -140,6 +140,18 @@ class IndexUtilities
...
+    /*!
+     * \return The spatial coordinate of the given side center.
+     *
+     * @param patch The patch on which the cell lives.
+     *
+     * @param side_idx The SideIndex describing the current side.
+     */
+    IBTK::VectorNd
+    getSideCenter(const SAMRAI::hier::Patch<NDIM> &patch,
+                  const SAMRAI::pdat::SideIndex<NDIM> &side_idx);
```
Comment (drwells) [inline ibtk/include/ibtk/IndexUtilities.h]: We should add a short test for this.

**Comment by drwells** [comment]
@amneetb Could you please send me the test code you mentioned you had for the non-rotating cylinder?

**Comment by amneetb** [comment]
> @amneetb Could you please send me the test code you mentioned you had for the non-rotating cylinder?

I need to rewrite it. I was hacking the NWT branch to do a quick-and-dirty check. If you can commit these changes to some branch, I can add the test.

**Comment by drwells** [comment]
Aside from the trivial merge conflict this is ready to go once we fix the examples: lets merge this once the testers are happy. @amneetb how should we fix the first two multiphase flow examples?

**Comment by amneetb** [comment]
> @amneetb how should we fix the first two multiphase flow examples?

I think we decided that we can pass an `std::ostream` argument to this function and that `main.cpp` should manage this stream object.

**Comment by drwells** [comment]
@amneetb Agreed; however, the issue with the two examples is that they use `IBHydrodynamicSurfaceForceEvaluator::computeHydrodynamicForce` which we removed.

**Comment by amneetb** [comment]
> `IBHydrodynamicSurfaceForceEvaluator::computeHydrodynamicForce` which we removed.

Oh this is easy. We just use `IBHydrodynamicSurfaceForceEvaluator::computeHydrodynamicForceTorque` in its place which is what I switched to in the NWT branch. Notice that we are already passing force and torque vectors in this function. We can just use this to populate an ostream file created in main.cpp. 

Something like 

`force_stream << time << pressure_force[0] << "\t" << pressure_force[1] << "\t"  << viscous_force[0] << "\t" << viscous_force[1] << std::endl; `

and 

`torque_stream << time << pressure_torque[2] << "\t"  << viscous_torque[2] << std::endl;`

This should be enough for these two-dimensional examples.

**Comment by amneetb** [comment]
> Something like
> 
> `force_stream << time << pressure_force[0] << "\t" << pressure_force[1] << "\t" << viscous_force[0] << "\t" << viscous_force[1] << std::endl; `
> 
> and
> 
> `torque_stream << time << pressure_torque[2] << "\t" << viscous_torque[2] << std::endl;`

We can also write just one file:

`force_torque_stream << time << pressure_force[0] << "\t" << pressure_force[1] << "\t" << viscous_force[0] << "\t" << viscous_force[1] << "\t" << pressure_torque[2] << "\t" << viscous_torque[2] << std::endl; `

**Comment by amneetb** [comment]
https://github.com/drwells/IBAMR/blob/brinkman-dependencies/src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp#L345

Note that we need to make sure cross product works in 2d also and does not error out. Torque is 3D even for 2D simulations (out of plane direction). I had IBTK::Vector3d instead of Vector2d for this reason in this routine.

**Comment by boyceg** [comment]
I was going to ask about that cross product routine in the implementation, which seems a bit odd. Could we just use 3D vectors in all cases, with the native cross product?

**Comment by amneetb** [comment]
Yes, please. I always stick to Vector3d for all of the rigid body dynamics code. It makes even physical sense because of torque.

**Comment by drwells** [comment]
I'll change that back. Shouldn't `X0` still be an N-dimensional vector, though?

**Comment by amneetb** [comment]
> I'll change that back. Shouldn't `X0` still be an N-dimensional vector, though?

Yes, but it not worth saving the memory of a double. You need `r` from `X0` which needs to be 2d vector for cross product. I would suggest sticking to vector3d for everything.

**Comment by drwells** [comment]
Easy enough.

**Comment by amneetb** [comment]
> @amneetb Could you please send me the test code you mentioned you had for the non-rotating cylinder?

Here you go:
[check_hydro_force.zip](https://github.com/IBAMR/IBAMR/files/3312427/check_hydro_force.zip)

It sets a linear pressure field with slope -1, i.e, increases downwards. Net pressure force on the body (cylinder in this case) i

`{Px, Py} = integral  [-grad(p) dv] = (0, volume of body)`

**Comment by drwells** [comment]
Ok, that should fix:
1. all 2D vs 3D issues
2. merge conflicts
3. test addition

So this is ready!

**Comment by amneetb** [comment]
Thanks!

PS: This block could be avoided 

https://github.com/drwells/IBAMR/blob/brinkman-dependencies/src/IB/IBHydrodynamicSurfaceForceEvaluator.cpp#L289-L292 

if we initialize `IBTK::Vector3d r_vec = IBTK::Vector3d::Zero()`

PPS: We can also add another line of check in the code for pressure torque. It should be zero analytically, as pressure forces pass through the COM of the cylinder.

**Comment by drwells** [comment]
Good call. I did both.

**Comment by amneetb** [comment]
Thanks! Next stop: `BrinkmanPenalizationRigidBodyDynamics.{h,cpp}`

**Comment by drwells** [comment]
Since @boyceg approved and @amneetb is happy lets merge.

**Pull request merged**